<!-- EKF NODE + NAVSAT NODE + EKF NODE-->
<!-- imu e odom from gazebo (re-published from publish_test_node to adjoint frame_id),
      gps from publish_test_node -->
<launch>

<!-- EKF WITH ODOM AND IMU -->
  <node pkg="nostop_robot_localization" type="ekf_localization_node" name="ekf_local" clear_params="true">
     <param name="frequency" value="30"/>
     <param name="publish_transform" value="true"/><!-- param adjoint by Niko -->
     <param name="sensor_timeout" value="0.1"/>
     <param name="two_d_mode" value="true"/>
      <param name="map_frame" value="SRworld"/>
      <param name="odom_frame" value="red_blue/odom"/>
       <param name="base_link_frame" value="red_blue/base_link"/>
       <param name="world_frame" value="red_blue/odom"/>
      <param name="transform_time_offset" value="0.0"/>
     <param name="odom0" value="/test_odom"/>
      <param name="imu0" value="/test_imu"/>
      <rosparam param="odom0_config">[true, true, false, 
                                false, false, true, 
                                true, false, false, 
                                false, false, true,
                                false, false, false]</rosparam>

      <rosparam param="imu0_config">[false, false, false, 
                               false, false, true, 
                               false, false, false, 
                               false, false, true,
                               true, false, false]</rosparam>
      <param name="odom0_differential" value="false"/>
      <param name="imu0_differential" value="false"/>
      <param name="odom0_relative" value="false"/>
      <param name="imu0_relative" value="false"/>
      <param name="imu0_remove_gravitational_acceleration" value="false"/>
      <param name="print_diagnostics" value="true"/>
     <rosparam param="process_noise_covariance">[0.05, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0.05, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0.06, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0.03, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0.03, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0.06, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0.025, 0,     0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0.025, 0,    0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0.04, 0,    0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0.01, 0,    0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.01, 0,    0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.02, 0,    0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.01, 0,    0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.01, 0,
                                                  0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.015]</rosparam>

           <rosparam param="initial_estimate_covariance">[1e-9, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    1e-9, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    1e-9, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-9, 0,    0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-9, 0,
                                                          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-9]</rosparam>


      <!--  Placeholder for output topic remapping-->
      <remap from="odometry/filtered" to="/output_of_ekf1"/>
    </node>
    
<!-- NAVSAT TRANSFORM NODE -->
<!--     <node pkg="nostop_robot_localization" type="navsat_transform_node" name="navsat_node" respawn="true"> -->
<!--       <param name="magnetic_declination_radians" value="2.38"/> -->
<!--       <param name="yaw_offset" value="0"/> -->
<!--       <param name="delay" value="0"/> -->
<!--       <param name="map_frame" value="SRworld"/> -->
<!--       <param name="odom_frame" value="nome_robot/odom"/> -->
<!--       <param name="base_link_frame" value="nome_robot/base_link"/> -->
<!--       <param name="world_frame" value="SRworld"/> -->
<!--       <param name="debug" value="true"/> -->
       <!--  Placeholder for input topic remapping-->
       <!--remap from="don't touch" to="my_topic" /-->
<!--       <remap from="/imu/data" to="/test_imu" /> -->
<!--       <remap from="/gps/fix" to="/test_gps" /> -->
<!--       <remap from="/odometry/filtered" to="/output_of_ekf1" /> -->
      
       <!-- Placeholders for output remapping. -->
<!--       <remap from="/odometry/gps" to="/odometry_from_gps"/> -->
<!--       <remap from="/gps/filtered" to="/gps_filtered"/>  -->
<!--     </node> -->

    
<!-- EKF WITH GPS_ODOM, ODOM AND IMU -->
<!--   <node pkg="nostop_robot_localization" type="ekf_localization_node" name="ekf_global" clear_params="true"> -->
<!--      <param name="frequency" value="30"/> -->
<!--      <param name="sensor_timeout" value="0.1"/> -->
<!--      <param name="two_d_mode" value="true"/> -->
<!--      <param name="map_frame" value="SRworld"/> -->
<!--      <param name="odom_frame" value="nome_robot/odom"/> -->
<!--      <param name="base_link_frame" value="nome_robot/base_link"/> -->
<!--      <param name="world_frame" value="SRworld"/> -->
<!--      <param name="transform_time_offset" value="0.0"/> -->
<!--      <param name="odom0" value="/test_odom"/> -->
<!--      <param name="odom1" value="/odometry_from_gps"/>   -->
<!--      <param name="imu0" value="/test_imu"/> -->
<!--      <rosparam param="odom0_config">[true, true, false,  -->
<!--                                 false, false, true,  -->
<!--                                 true, false, false,  -->
<!--                                 false, false, true, -->
<!--                                 false, false, false]</rosparam> -->
<!--                                  -->
<!--       <rosparam param="odom1_config">[true, true, false,  -->
<!--                                false, false, false,  -->
<!--                                false, false, false,  -->
<!--                                false, false, false, -->
<!--                                false, false, false]</rosparam> -->
<!--  -->
<!--   -->

<!--       <rosparam param="imu0_config">[false, false, false,  -->
<!--                                false, false, true,  -->
<!--                                false, false, false,  -->
<!--                                false, false, true, -->
<!--                                true, false, false]</rosparam> -->
<!--     <param name="odom0_differential" value="false"/> -->
<!--       <param name="odom1_differential" value="false"/> -->
<!--       <param name="imu0_differential" value="false"/> -->
<!--       <param name="odom0_relative" value="false"/> -->
<!--       <param name="imu0_relative" value="false"/> -->
<!--       <param name="imu0_remove_gravitational_acceleration" value="false"/> -->
<!--      <rosparam param="process_noise_covariance">[0.05, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0.05, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0.06, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0.03, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0.03, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0.06, 0,     0,     0,    0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0.025, 0,     0,    0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0,     0.025, 0,    0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0,     0,     0.04, 0,    0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0,     0,     0,    0.01, 0,    0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.01, 0,    0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.02, 0,    0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.01, 0,    0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.01, 0, -->
<!--                                                   0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.015]</rosparam> -->

<!--           <rosparam param="initial_estimate_covariance">[1e-9, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    1e-9, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    1e-9, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-9, 0,    0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-9, 0, -->
<!--                                                           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-9]</rosparam> -->

<!--       <remap from="/odometry/filtered" to="/nome_robot/localizer/odometry/final"/> -->
      
<!--       </node> -->
    
</launch>
